egressgw: set ifindex for egress interface in IPv6 policy entry

julianwiedmann opened the pull request:
Add support for pushing down the ifindex of the egress interface (https://github.com/cilium/cilium/pull/36151).

This avoids the usability annoyance of adding custom routes on the gateway node, as described in https://github.com/cilium/cilium/pull/30488.


```release-note
The egress interface for IPv6 traffic matched by an Egress Gateway policy is entirely determined by the `egressGateway` section in the policy, using either the selected `interface` or the interface with the default route. When an `egressIP` is specified, the interface continues to be selected according to the configured IP routing on the node.
```

julianwiedmann committed:
egressgw: datapath support for ifindex in IPv6 policy entry

Enable the EGW datapath on the gateway node to obtain the ifindex of the
egress interface from the EGW policy entry, instead of needing a FIB
lookup.

No change in behavior, as the agent still pushes down 0 as the ifindex
value (see the change in updateEgressRules6()). The needed controlplane
changes follow in a subsequent patch.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

rgo3 commented on bpf/lib/common.h:
```diff
@@ -472,6 +472,7 @@ struct egress_gw_policy_entry6 {
 	union v6addr egress_ip;
 	__u32 gateway_ip;
 	__u32 reserved[3]; /* reserved for future extension, e.g. v6 gateway_ip */
+	__u32 egress_ifindex;
```
Drive by comment: was this u32 variable one of the three u32 we reserved with the array above or do we still need 96 bits reserved for future use?

julianwiedmann replied:
I liked your idea of eventually turning those unused bytes into an IPv6 address (for IPv6-only configs).

rgo3 replied:
Just realizing we'll need more bits reserved anyway. In my initial PR it should have been 5x32 bits. 128 for the ipv6 gateway IP, and 32 for a future use-case you had in mind, which is presumably this one?

julianwiedmann replied:
> In my initial PR it should have been 5x32 bits. 128 for the ipv6 gateway IP

I think we can follow Paul's approach in https://github.com/cilium/cilium/pull/38296, and unionize the IPv4 and IPv6 address field.

> and 32 for a future use-case you had in mind, which is presumably this one?

yep!

rgo3 replied:
>I think we can follow Paul's approach in https://github.com/cilium/cilium/pull/38296, and unionize the IPv4 and IPv6 address field.

Oh, that's neat. Then 3 reserved is still correct, phew.

julianwiedmann committed:
egressgw: controlplane support for egress ifindex in IPv6 policy entry

For CEGPs that select the egress interface by name (`interface` parameter)
or by default-route, also push the interface ifindex down into the
datapath. This avoids custom routing setup for forwarding EGW traffic to
the correct network interface.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>
